db/view/view.cc

```diff
@@ -2645,6 +2646,30 @@ update_backlog node_update_backlog::fetch() {
     return std::max(fetch_shard(this_shard_id()), _max.load(std::memory_order_relaxed));
 }
 
+future<std::optional<update_backlog>> node_update_backlog::fetch_if_changed(std::optional<db::view::update_backlog> last_backlog) {
+    auto np = co_await map_reduce(boost::irange(0u, smp::count),
+            [this] (shard_id shard) {
+                return smp::submit_to(shard, [this, shard] {
+                    if (_backlogs[shard].need_publishing) {
+                        _backlogs[shard].need_publishing = need_publishing::no;
+                        return need_publishing::yes;
+                    }
+                    return need_publishing::no;
+                });
+            },
+            need_publishing::no,
+            [] (need_publishing a, need_publishing b) {
+                return a || b;
+            });
+
+    auto max = fetch();
+    if (!np && last_backlog && *last_backlog == max) {
```

Comment:
I asked about this before, but the comment was resolved.

Isn't it sufficient to just check:

```c++
if (!np) {
```

?

Previously, we used to compare the last backlog vs the current one in order to decide whether it needs publishing. Now, we have the `need_publishing` flag. Is the `last_backlog && *last_backlog == max` still necessary?

Reply:
The whole if..else could be written as a ternary operator:

```c++
return np ? max : std::nullopt;
```

I didn't try to compile it, there is a chance that you'll get a type error because both branches don't have the same type, however it would be nice to try it.

Reply:
> I asked about this before, but the comment was resolved.
> 
> Isn't it sufficient to just check:
> 
> ```c++
> if (!np) {
> ```
> 
> ?
> 
> Previously, we used to compare the last backlog vs the current one in order to decide whether it needs publishing. Now, we have the `need_publishing` flag. Is the `last_backlog && *last_backlog == max` still necessary?

After https://github.com/scylladb/scylladb/pull/18804 is merged and the `need_publishing` is also updated in that context, the `last_backlog && *last_backlog == max` check won't be needed anymore. Without #18804 the `need_publishing` flag is not set when a backlog drops, so we have no other way of detecting it other than comparing with  the last sent backlog

Reply:
I used this in the last rebase